internal/pkg/api/handleCheckin.go

```diff
@@ -495,84 +493,60 @@ func (ct *CheckinT) processUpgradeDetails(ctx context.Context, agent *model.Agen
+			return nil, fmt.Errorf("%w %s: %w", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGFAILED, err)
 		}
 		if meta.ErrorMsg == "" {
 			vSpan.End()
-			return fmt.Errorf("%w: %s metadata contains empty error_msg attribute", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGFAILED)
+			return nil, fmt.Errorf("%w: %s metadata contains empty error_msg attribute", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGFAILED)
 		}
 		// Repack metadata in failed case as the agent may send UPG_DOWNLOADING attributes.
 		if err = details.Metadata.FromUpgradeMetadataFailed(meta); err != nil {
 			vSpan.End()
-			return fmt.Errorf("%w %s: unable to repack metadata: %w", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGFAILED, err)
+			return nil, fmt.Errorf("%w %s: unable to repack metadata: %w", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGFAILED, err)
 		}
 	case UpgradeDetailsStateUPGSCHEDULED:
 		if details.Metadata == nil {
 			vSpan.End()
-			return fmt.Errorf("%w: metadata missing", ErrInvalidUpgradeMetadata)
+			return nil, fmt.Errorf("%w: metadata missing", ErrInvalidUpgradeMetadata)
 		}
 		meta, err := details.Metadata.AsUpgradeMetadataScheduled()
 		if err != nil {
 			vSpan.End()
-			return fmt.Errorf("%w %s: %w", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGSCHEDULED, err)
+			return nil, fmt.Errorf("%w %s: %w", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGSCHEDULED, err)
 		}
 		if meta.ScheduledAt.IsZero() {
 			vSpan.End()
-			return fmt.Errorf("%w: %s metadata contains empty scheduled_at attribute", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGSCHEDULED)
+			return nil, fmt.Errorf("%w: %s metadata contains empty scheduled_at attribute", ErrInvalidUpgradeMetadata, UpgradeDetailsStateUPGSCHEDULED)
 
 		}
 	default:
 	}
 	vSpan.End()
 
-	doc := bulk.UpdateFields{
-		dl.FieldUpgradeDetails: details,
-	}
-	if agent.UpgradeAttempts != nil && details.State == UpgradeDetailsStateUPGWATCHING {
-		doc[dl.FieldUpgradeAttempts] = nil
-	}
-
-	body, err := doc.Marshal()
+	detailsBody, err := json.Marshal(details)
 	if err != nil {
-		return err
-	}
-	return ct.bulker.Update(ctx, dl.FleetAgents, agent.Id, body, bulk.WithRefresh(), bulk.WithRetryOnConflict(3))
-}
-
-func (ct *CheckinT) markUpgradeComplete(ctx context.Context, agent *model.Agent) error {
-	// nop if there are no checkin details, and the agent has no details
-	if agent.UpgradeDetails == nil {
-		return nil
+		return nil, fmt.Errorf("failed to marshal upgrade details: %w", err)
 	}
-	span, ctx := apm.StartSpan(ctx, "Mark update complete", "update")
-	span.Context.SetLabel("agent_id", agent.Agent.ID)
-	defer span.End()
-	// if the checkin had no details, but agent has details treat like a successful upgrade
-	doc := bulk.UpdateFields{
-		dl.FieldUpgradeDetails:   nil,
-		dl.FieldUpgradeStartedAt: nil,
-		dl.FieldUpgradeStatus:    nil,
-		dl.FieldUpgradedAt:       time.Now().UTC().Format(time.RFC3339),
```

Comment:
I am probably missing something but I don't see the  equivalent of setting `upgradedAt` field in the new version of `processUpgradeDetails`. Is this handled elsewhere ?

Reply:
Yes that is automatically handled here https://github.com/elastic/fleet-server/pull/5834/files#diff-945b012d5b50c9ece3bce4550dfaa563337e1b905e1b69ed8d28486d39e2b832R178